code/hud/hudtarget.cpp

```diff
@@ -3979,46 +3970,117 @@ void HudGaugeLeadIndicator::renderLeadCurrentTarget()
 		dist_to_target = hud_find_target_distance(targetp, Player_obj);
 	}
 
-	srange = ship_get_secondary_weapon_range(Player_ship);
+	//Moved here to avoid potentially leaving wip undefined. These values will be
+	//overwritten if using non-default leadIndicatorBehaviors, this is desired.
+	bank_to_fire = hud_get_best_primary_bank(&prange);
+	wip = &Weapon_info[swp->primary_bank_weapons[bank_to_fire]];
 
-	if ( (swp->current_secondary_bank >= 0) && (swp->secondary_bank_weapons[swp->current_secondary_bank] >= 0) )
-	{
-		int bank = swp->current_secondary_bank;
-		tmp = &Weapon_info[swp->secondary_bank_weapons[bank]];
-		if ( !(tmp->is_homing()) && !(tmp->is_locked_homing() && Player->target_in_lock_cone) ) {
-			//The secondary lead indicator is handled farther below if it is a non-locking type
-			srange = -1.0f;
+	if (Lead_indicator_behavior != leadIndicatorBehavior::DEFAULT && Player_ship->flags[Ship::Ship_Flags::Primary_linked]) {
+		vec3d averaged_lead_pos;
+		vm_vec_zero(&averaged_lead_pos);
+		int average_instances = 0;
+		for (int i = 0; i < swp->num_primary_banks; i++) {
+			auto bank = i;
+			// calculate the range of the weapon, and only display the lead target indicator
+			// if the weapon can actually hit the target
+			Assert(bank >= 0 && bank < swp->num_primary_banks);
+			Assert(swp->primary_bank_weapons[bank] < weapon_info_size());
+
+			if (swp->primary_bank_weapons[bank] < 0)
+				continue;
+
+			wip = &Weapon_info[swp->primary_bank_weapons[bank]];
+
+			if (wip->wi_flags[Weapon::Info_Flags::Nolink] ||  (wip->wi_flags[Weapon::Info_Flags::Ballistic] && swp->primary_bank_ammo[bank] <= 0))
+				continue;
+
+			if (bank < 0)
+				continue;
```

Comment:
If you have the bank >= 0 assert further up, this check here makes little sense. Either remove it, or at least move it far enough up so that bank isn't already used to index into arrays